### PR TITLE
feat(connect-common): modify storage

### PR DESCRIPTION
**Pull request by tomasklim**
## Description

Redone connect storage a little bit to be able to use it for analytics.

## Related Issue

https://github.com/trezor/trezor-suite/issues/5397

### CODE REVIEW

**Comment by Nodonisko** [comment]
I am not sure how this is used in connect, but we don't have localstorage available in mobile app so I afraid it could break it. Also what about node env?

**Comment by mroz22** [comment]
We discussed that yesterday, in the end, we want to have just some kind of abstract facade API that will only receive "whatever storage interface" from outside. We are actually alrady facing the problem that like this it does not work in node https://github.com/trezor/trezor-suite/actions/runs/3289271974/jobs/5420714331#step:7:2055

File: packages/connect-common/src/storage.ts [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    origin?: string;
+    device?: string;
+}
+
+export interface Store {
+    [BROWSER_KEY]?: boolean;
+    [PERMISSIONS_KEY]?: Permission[];
+}
+
+type SaveCallback = (state: Store, setState: (newState: Store) => void) => void;
```
Comment (dahaca) [inline packages/connect-common/src/storage.ts]:
```suggestion
type GetNewStateCallback = (currentState: Store) => Store;
```

File: packages/connect-common/src/storage.ts [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+export const save = (saveCallback: SaveCallback, temporary = false) => {
+    if (temporary) {
+        saveCallback(memoryStorage, store => {
+            memoryStorage = store;
+        });
+        return;
+    }
+    const storage = getPermanentStorage();
+    saveCallback(storage, store => localStorage.setItem(storageName, JSON.stringify(store)));
+};
```
Comment (dahaca) [inline packages/connect-common/src/storage.ts]:
```suggestion
export const save = (getNewState: GetNewStateCallback, temporary = false) => {
   if (temporary) {
       memoryStorage  = getNewState(memoryStorage);
             
       return;
    }
    
    const newState = getNewState(getPermanentStorage());
    localStorage.setItem(storageName, JSON.stringify(newState))
};
```

File: packages/connect-popup/src/view/browser.ts [hunk 3]
```diff
@@ -47,7 +47,12 @@ export const initBrowserView = (validation = true) => {
 
     ackButton.onclick = () => {
         if (rememberCheckbox && rememberCheckbox.checked) {
-            storage.save(storage.BROWSER_KEY, true);
+            storage.save((state, setState) => {
+                setState({
+                    ...state,
+                    [storage.BROWSER_KEY]: true,
+                });
+            });
```
Comment (dahaca) [inline packages/connect-popup/src/view/browser.ts]:
```suggestion
            storage.save((state) => ({
                    ...state,
                    [storage.BROWSER_KEY]: true,
            }));
```

**Review by dahaca — COMMENTED** [review]
this API could be a bit clearer, what do you say?

File: packages/connect-common/src/storage.ts (hunk 2 above)
Comment (tomasklim) [reply packages/connect-common/src/storage.ts]: done

File: packages/connect-popup/src/view/browser.ts (hunk 3 above)
Comment (tomasklim) [reply packages/connect-popup/src/view/browser.ts]: done

**Pull request closed without merging**